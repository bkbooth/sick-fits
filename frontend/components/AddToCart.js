import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART($itemId: ID!) {
    addToCart(itemId: $itemId) {
      id
      quantity
    }
  }
`;

const AddToCart = ({ itemId }) => (
  <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ itemId }}>
    {(addToCart, { loading }) => (
      <button onClick={addToCart} disabled={loading}>
        ðŸ›’ Add{loading ? 'ing' : ''} to cart
      </button>
    )}
  </Mutation>
);

AddToCart.propTypes = {
  itemId: PropTypes.string.isRequired,
};

export default AddToCart;
