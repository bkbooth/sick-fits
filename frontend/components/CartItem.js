import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import RemoveFromCart from './RemoveFromCart'
import formatMoney from '../lib/formatMoney'

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`

const CartItem = ({ cartItem }) => {
  if (!cartItem.item) return (
    <CartItemStyles>
      <p>This item has been removed.</p>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  )
  return <CartItemStyles>
    <img src={cartItem.item.image} width="100" alt={cartItem.item.title} />
    <div className="cart-item-details">
      <h3>{cartItem.item.title}</h3>
      <p>
        {formatMoney(cartItem.item.price * cartItem.quantity)}
        {' - '}
        <em>{cartItem.quantity} Ã— {formatMoney(cartItem.item.price)} each</em>
      </p>
    </div>
    <RemoveFromCart id={cartItem.id} />
  </CartItemStyles>
}

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    id: PropTypes.string,
    quantity: PropTypes.number,
    item: PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      description: PropTypes.string,
      image: PropTypes.string,
      price: PropTypes.number,
    })
  }).isRequired,
}

export default CartItem
