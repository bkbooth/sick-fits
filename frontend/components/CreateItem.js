import React, { Component } from 'react'
import Router from 'next/router'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  }

  handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', 'sick-fits')

    const res = await fetch('https://api.cloudinary.com/v1_1/bkbooth/image/upload', {
      method: 'POST',
      body: data,
    })
    const file = await res.json()

    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    })
  }

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
              onSubmit={async e => {
              // Stop the form from submitting
              e.preventDefault()
              // Call the mutation
              const res = await createItem()
              // Redirect user to the single item page
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
            data-test="form"
          >
            <Error error={error} />

            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
                {this.state.image && <img src={this.state.image} width="200" alt="Upload Preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a description"
                  required
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
