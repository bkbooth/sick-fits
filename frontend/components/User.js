import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          title
          price
          image
        }
      }
    }
  }
`;

const User = ({ children }) => {
  const payload = useQuery(CURRENT_USER_QUERY);
  return children(payload);
};

User.propTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
