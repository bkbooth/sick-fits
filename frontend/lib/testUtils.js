import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

export const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

export const fakeItem = () => ({
  __typename: 'Item',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  largeImage: `${casual.word}-large.jpg`,
  user: null,
});

export const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: casual.uuid,
  quantity: casual.integer(1, 5),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

export const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  quantity: casual.integer(1, 5),
});

export const fakeOrder = () => ({
  __typename: 'Order',
  id: casual.uuid,
  charge: `ch_${casual.uuid.split('-')[4]}`,
  total: casual.integer(10000, 100000),
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: casual.date('YYYY-MM-DD[T]HH:mm:ss[Z]'),
  user: fakeUser(),
});

export function typeInto(wrapper, { inputType = 'input', name, type = 'text', value }) {
  wrapper
    .find(`${inputType}[name="${name}"]`)
    .simulate('change', { target: { name, type, value } });
}
