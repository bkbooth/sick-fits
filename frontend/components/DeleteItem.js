import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { ALL_ITEMS_QUERY } from 'components/Items';

export const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM($itemId: ID!) {
    deleteItem(id: $itemId) {
      id
    }
  }
`;

const DeleteItem = ({ itemId }) => {
  function createDeleteHandler(mutation) {
    return () =>
      confirm('Are you sure you want to delete this item?') &&
      mutation().catch(error => alert(error.message));
  }

  function cacheUpdate(cache, payload) {
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  }

  return (
    <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ itemId }} update={cacheUpdate}>
      {(deleteItem, { loading }) => (
        <button onClick={createDeleteHandler(deleteItem)} disabled={loading}>
          🗑️ Delet{loading ? 'ing' : 'e'} item
        </button>
      )}
    </Mutation>
  );
};

DeleteItem.propTypes = {
  itemId: PropTypes.string.isRequired,
};

export default DeleteItem;
