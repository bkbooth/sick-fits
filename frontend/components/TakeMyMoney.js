import React from 'react';
import { useMutation } from '@apollo/react-hooks';
import StripeCheckout from 'react-stripe-checkout';
import gql from 'graphql-tag';
import Router from 'next/router';
import NProgress from 'nprogress';
import calculateTotalPrice from 'lib/calculateTotalPrice';
import countCartItems from 'lib/countCartItems';
import User, { CURRENT_USER_QUERY } from 'components/User';

export const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER($token: String!) {
    createOrder(token: $token) {
      id
    }
  }
`;

const TakeMyMoney = ({ children }) => {
  const [createOrder] = useMutation(CREATE_ORDER_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function onToken({ id }) {
    NProgress.start();
    try {
      const { data } = await createOrder({ variables: { token: id } });
      Router.push('/orders/[orderId]', `/orders/${data.createOrder.id}`);
    } catch (error) {
      NProgress.done();
      alert(error.message);
    }
  }

  return (
    <User>
      {({ data: { me }, loading }) => {
        if (loading) return null;
        return (
          <StripeCheckout
            name="Sick Fits!"
            description={`Order of ${countCartItems(me.cart)} items`}
            image={me.cart.length > 0 && me.cart[0].item ? me.cart[0].item.image : null}
            amount={calculateTotalPrice(me.cart)}
            currency="AUD"
            stripeKey="pk_test_nmnQGOikwUBoSjWEFR2tvQdR"
            email={me.email}
            token={onToken}
          >
            {children}
          </StripeCheckout>
        );
      }}
    </User>
  );
};

export default TakeMyMoney;
