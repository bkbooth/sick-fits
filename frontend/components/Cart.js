import React from 'react'
import { Query, Mutation } from 'react-apollo'
import { adopt } from 'react-adopt'
import gql from 'graphql-tag'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'
import User from './User'
import CartItem from './CartItem'
import TakeMyMoney from './TakeMyMoney'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'
import apostrophiseName from '../lib/apostrophiseName'

const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`

const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
})

const Cart = () => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const me = user.data.me
      if (!me) return null
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton
              title="close"
              onClick={toggleCart}
            >Ã—</CloseButton>
            <Supreme>{apostrophiseName(me.name)} Cart</Supreme>
            <p>You have {me.cart.length} item{me.cart.length === 1 ? '': 's'} in your cart.</p>
          </header>

          <ul>
            {me.cart.map(cartItem => <CartItem cartItem={cartItem} key={cartItem.id} />)}
          </ul>

          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <TakeMyMoney>
              <SickButton disabled={!me.cart.length}>Checkout</SickButton>
            </TakeMyMoney>
          </footer>
        </CartStyles>
      )
    }}
  </Composed>
)

export default Cart
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION }
