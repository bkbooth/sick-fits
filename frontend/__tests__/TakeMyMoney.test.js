import { ApolloConsumer } from '@apollo/react-common';
import { MockedProvider, wait } from '@apollo/react-testing';
import { mount } from 'enzyme';
import Router from 'next/router';
import NProgress from 'nprogress';
import { act } from 'react-dom/test-utils';
import { fakeCartItem, fakeOrder, fakeUser } from 'lib/testUtils';
import TakeMyMoney, { CREATE_ORDER_MUTATION } from 'components/TakeMyMoney';
import { CURRENT_USER_QUERY } from 'components/User';

const me = fakeUser();
const cartItem = fakeCartItem();
const order = fakeOrder();
const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: { ...me, cart: [cartItem] } } },
  },
  {
    request: { query: CREATE_ORDER_MUTATION, variables: { token: 'tok_abc123' } },
    result: { data: { createOrder: order } },
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: { ...me, orders: [order] } } },
  },
];

async function render() {
  const wrapper = mount(
    <MockedProvider mocks={mocks}>
      <TakeMyMoney>
        <button>Checkout</button>
      </TakeMyMoney>
    </MockedProvider>
  );
  await act(() => wait());
  wrapper.update();
  return wrapper;
}

describe('<TakeMyMoney />', () => {
  beforeEach(() => {
    Router.push = jest.fn();
    NProgress.start = jest.fn();
    NProgress.done = jest.fn();
  });

  afterEach(() => {
    Router.push.mockReset();
    NProgress.start.mockReset();
    NProgress.done.mockReset();
  });

  it('renders and matches snapshot', async () => {
    const wrapper = await render();
    expect(wrapper.find('ReactStripeCheckout')).toMatchSnapshot();
  });

  it('routes to the order page after order is created', async () => {
    const wrapper = await render();
    await act(async () => {
      wrapper.find('ReactStripeCheckout').prop('token')({ id: 'tok_abc123' });
      await wait();
    });
    expect(Router.push).toHaveBeenCalledTimes(1);
    expect(Router.push).toHaveBeenCalledWith('/orders/[orderId]', `/orders/${order.id}`);
  });

  it('shows the progress bar while creating order', async () => {
    const wrapper = await render();
    await act(async () => {
      wrapper.find('ReactStripeCheckout').prop('token')({ id: 'tok_abc123' });
      await wait();
    });
    expect(NProgress.start).toHaveBeenCalledTimes(1);
  });
});
