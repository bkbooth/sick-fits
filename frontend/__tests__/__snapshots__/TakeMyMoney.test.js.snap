// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TakeMyMoney /> renders and matches snapshot 1`] = `
<ReactStripeCheckout
  ComponentClass="span"
  amount={117545}
  className="StripeCheckout"
  currency="AUD"
  description="Order of 5 items"
  email="Hills.Princess@Vena.us"
  image="fugiat.jpg"
  label="Pay With Card"
  locale="auto"
  name="Sick Fits!"
  reconfigureOnUpdate={false}
  stripeKey="pk_test_nmnQGOikwUBoSjWEFR2tvQdR"
  token={[Function]}
  triggerEvent="onClick"
>
  <span
    onClick={[Function]}
  >
    <button>
      Checkout
    </button>
  </span>
</ReactStripeCheckout>
`;
